Refactor sample tests configuration to reduce the efforts of adding samples.

numerology opened the pull request:
Refactor sample tests configuration to reduce the efforts of adding samples.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/1730)
<!-- Reviewable:end -->

numerology commented:
/test kubeflow-pipeline-sample-test

numerology commented:
Will follow up with another PR to solve following AIs:
1. Move repeated environment setup into one place to speed up test.
2. Move test-specific environment setup into .py or .ipynb to keep test infra clean.
3. Further naming convention issue.

PTAL. Thanks!

Ark-kun commented on test/sample-test/check_notebook_results.py:
```diff
@@ -18,46 +18,49 @@
-                      type=str,
-                      required=True,
-                      help='The exit code of the bash command that runs the test.')
+  parser.add_argument('--experiment', type=str, help='The experiment name')
```
Are there any actual changes in this files?

We usually separate refactoring or restyling PRs from change PRs. This way the diff is much smaller and the reviewers can focus on the actual change.

numerology replied:
Thanks for your suggestion. Reverted this change.

gaoning777 commented:
Could you also update the links in the READMEs to make sure they are not broken. For example, tfx-cab-classification/README.md has a broken link.

numerology commented:
@gaoning777 Thanks for your good catch! Have checked the link in the README.md and fixed it. Note that there is an ongoing PR #1751 also trying to fix some of the links.

numerology commented:
> You have suggested to keep the naming consistent, which I think is a great idea.
> Why don't we convert the names using underscore as the separator for the words in the file/dir names?
> For example, kubeflow_training_classification. Since the majority of the sample directory names use the underscore. WDYT?

Agreed. Done.

gaoning777 commented on test/sample-test/run_sample_test.py:
```diff
@@ -0,0 +1,190 @@
+# Copyright 2019 Google LLC
```
Is this going to be the entrypoint, replacing the run_test.sh?
If so, please also update the dockerfile test/sample-test/Dockerfile.

gaoning777 replied:
Never mind. I thought the run_test.sh is deleted.

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
-  cd ${BASE_DIR}/samples/core/tfx
-
-  dsl-compile --py taxi-cab-classification-pipeline.py --output taxi-cab-classification-pipeline.yaml
+  gsutil cp ${SAMPLE_TEST_RESULT} ${RESULTS_GCS_DIR}/${SAMPLE_TEST_OUTPUT}
```
Should this be
gsutil cp ${SAMPLE_TEST_RESULT} ${RESULTS_GCS_DIR}/${SAMPLE_TEST_RESULT}
?

numerology replied:
Good catch. Thanks!

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
+################################################################################
+check_notebook_result() {
+  jupyter nbconvert --to python $1.ipynb
+  pip3 install tensorflow==1.8.0
```
This is not required by all notebook samples, maybe we can move it to the notebook sample file instead of in the test infra?

numerology replied:
Yes we discussed about it. I think it's better to do that in another PR since this one is already big enough. Do you agree?

gaoning777 replied:
That's totally fine. How about add a TODO comment to keep track?

numerology replied:
Added TODO.

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
+  cd ${TEST_DIR}
+
+  if [[ $1 != "dsl_static_type_checking" ]]; then
+    # If the sample does not require pipeline running under a experiment, put it
```
this 'if' is when the sample does require pipeline running in an experiment, correct?

numerology replied:
Exactly.

gaoning777 replied:
which means the comments are incorrect?

gaoning777 replied:
Feel free to include this in another PR since the tests are taking lots of time.

numerology replied:
Oh sorry I missed that. Good catch!

numerology replied:
> Feel free to include this in another PR since the tests are taking lots of time.

That's okay. PR should be separated by their purposes, I think.

gaoning777 commented:
The component yaml(https://github.com/kubeflow/pipelines/blob/master/test/component_test.yaml) might also need to be updated since the test-name is updated.

gaoning777 commented:
> Will follow up with another PR to solve following AIs:
> 
> 1. Move repeated environment setup into one place to speed up test.
> 2. Move test-specific environment setup into .py or .ipynb to keep test infra clean.
> 3. Further naming convention issue.
> 
> PTAL. Thanks!

Can I kindly request you to add these in the codes as comments starting with TODO. 
e.g.
a=hello
# TODO: declare b.

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
+  #TODO(numerology): convert the sed commands to sed -e
+  # 's|gcr.io/ml-pipeline/|gcr.io/ml-pipeline-test/' and tag replacement. Also
+  # let the postsubmit tests refer to yaml files.
+
```
Could we keep the sed command for the kubeflow_training_classification as well?

gaoning777 replied:
https://github.com/kubeflow/pipelines/blob/1407d86497bc0de3e4a20d5a6bb9689dc6085346/test/sample-test/run_test.sh#L143

numerology replied:
Agreed. Done.

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
   DEPLOYER_MODEL=`cat /proc/sys/kernel/random/uuid`
   DEPLOYER_MODEL=Notebook_tfx_taxi_`echo ${DEPLOYER_MODEL//-/_}`
 
   cd ${BASE_DIR}/samples/core/kubeflow_pipeline_using_TFX_OSS_components
```
Isn't this already in the sample folder after running the 'prepare' function above?

numerology replied:
Good catch. Thanks!

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
-      -p BASE_IMAGE ${TARGET_IMAGE_PREFIX}pusherbase:dev -p TARGET_IMAGE ${TARGET_IMAGE_PREFIX}pusher:dev -p TARGET_IMAGE_TWO ${TARGET_IMAGE_PREFIX}pusher_two:dev \
+  if [[ -n "${DATAFLOW_TFT_IMAGE}" ]]; then
+    papermill --prepare-only -p EXPERIMENT_NAME "${TEST_NAME}-test" -p OUTPUT_DIR \
+    ${RESULTS_GCS_DIR} -p PROJECT_NAME ml-pipeline-test \
```
nit: the same indentation for better readability.

numerology replied:
Done. Thanks.

gaoning777 commented on test/sample-test/run_test.sh:
```diff
@@ -101,242 +103,217 @@ while [ "$1" != "" ]; do
-    papermill --prepare-only -p EXPERIMENT_NAME notebook-tfx-test -p OUTPUT_DIR ${RESULTS_GCS_DIR} -p PROJECT_NAME ml-pipeline-test \
-      -p BASE_IMAGE ${TARGET_IMAGE_PREFIX}pusherbase:dev -p TARGET_IMAGE ${TARGET_IMAGE_PREFIX}pusher:dev -p TARGET_IMAGE_TWO ${TARGET_IMAGE_PREFIX}pusher_two:dev \
+    papermill --prepare-only -p EXPERIMENT_NAME "${TEST_NAME}-test" -p \
+    OUTPUT_DIR ${RESULTS_GCS_DIR} -p PROJECT_NAME ml-pipeline-test \
```
the same as above: indentation

numerology replied:
Done.

gaoning777 commented:
With the comments deleted: python3 clean_cmle_models.py
Let's remove the file from the sample-test directory.

numerology commented:
Thanks for the review @gaoning777 !
- Fixed component_test.yaml
- Lint indent in run_test.sh
- Added TODO items
- Removed redundant command in run_test.sh

PTAL.

Ark-kun commented:
/lgtm

numerology commented:
/test kubeflow-pipeline-e2e-test

Ark-kun commented on test/sample-test/run_sample_test.py:
```diff
@@ -0,0 +1,190 @@
+  job_name = args.testname + '_sample'
+  ###### Test-specific parameters #######
+  if args.testname == 'tfx_cab_classification':
+    params = {
```
Nit: These are "arguments", not "parameters".
These are parameters:
```python
def my_func(param1, param2):
```

These are arguments (for the parameters):
```python
my_func(param1=arg1, param2=arg2):
```

numerology replied:
Thanks for correcting!

Ark-kun commented:
> 2\. Move test-specific environment setup into .py or .ipynb to keep test infra clean.

The test cases may be even moved from the script to individual test case files.
If we had full support for graph components, each test case would just be a task:

```yaml
component_ref:
    url: https://raw.githubusercontent.com/kubeflow/pipelines/master/samples/core/tfx/taxi-cab-classification-pipeline.component.yaml
arguments:
    project: ml-pipeline-test
    column-names: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/column-names.json
    evaluation: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/eval20.csv
    train: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/train50.csv
    hidden-layer-size: 5
    steps: 5
```

And maybe the whole test set would be a pipeline:
```yaml
name: TFX tests
graph:
    tasks:
        test-1:
            component_ref:
                url: https://raw.githubusercontent.com/kubeflow/pipelines/master/samples/core/tfx/taxi-cab-classification-pipeline.component.yaml
            arguments:
                project: ml-pipeline-test
                column-names: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/column-names.json
                evaluation: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/eval20.csv
                train: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/train50.csv
                hidden-layer-size: 5
                steps: 5

        test-2:
            ...
```

Ark-kun commented:
/lgtm

numerology commented:
> > 2. Move test-specific environment setup into .py or .ipynb to keep test infra clean.
> 
> The test cases may be even moved from the script to individual test case files.
> If we had full support for graph components, each test case would just be a task:
> 
> ```yaml
> component_ref:
>     url: https://raw.githubusercontent.com/kubeflow/pipelines/master/samples/core/tfx/taxi-cab-classification-pipeline.component.yaml
> arguments:
>     project: ml-pipeline-test
>     column-names: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/column-names.json
>     evaluation: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/eval20.csv
>     train: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/train50.csv
>     hidden-layer-size: 5
>     steps: 5
> ```
> 
> And maybe the whole test set would be a pipeline:
> 
> ```yaml
> name: TFX tests
> graph:
>     tasks:
>         test-1:
>             component_ref:
>                 url: https://raw.githubusercontent.com/kubeflow/pipelines/master/samples/core/tfx/taxi-cab-classification-pipeline.component.yaml
>             arguments:
>                 project: ml-pipeline-test
>                 column-names: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/column-names.json
>                 evaluation: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/eval20.csv
>                 train: gs://ml-pipeline-dataset/sample-test/taxi-cab-classification/train50.csv
>                 hidden-layer-size: 5
>                 steps: 5
> 
>         test-2:
>             ...
> ```

Thanks for the suggestion @Ark-kun . I will take that into consideration for further refactoring and future test design.

numerology commented:
/test kubeflow-pipeline-e2e-test

gaoning777 commented:
Thanks jiaxiao to improve our sample test infra and refactor the codes. This is really important to keep the product quality high and saves huge engineering efforts in the future to add more tests for samples.

numerology commented:
Thanks for your reviews and suggestions! @gaoning777 @Ark-kun

numerology commented:
/test kubeflow-pipeline-e2e-test

numerology commented:
Synced with head to fix e2e test. Please relgtm, thx!

gaoning777 commented:
/lgtm

numerology commented:
/test kubeflow-pipeline-e2e-test
